### PR TITLE
Support regex matching for source and sink filters

### PR DESCRIPTION
**Pull request by abishekvashok**
Adds support for regex matching for source name and kind and sink name
and kind filters. The regex matching occurs in the frontend and no
modification is hence required in the backend.

Signed-off-by: Abishek V Ashok <abishekvashok@gmail.com>
Fixes: https://github.com/MLH-Fellowship/pyre-check/issues/46

### CODE REVIEW

**Comment by abishekvashok** [comment]
After the changes:

Screenshot 2021-08-05 at 11 22 07 AM

File: sapp/ui/frontend/src/Filter.js [hunk 1]
```diff
@@ -211,17 +214,50 @@ const SourceNames = (
...
+              value={props.currentFilter.source_names}
+              options={allSourceNames.map(value => ({value}))}
+              style={{width: '100%'}}
+              onChange={source_names =>
+                props.setCurrentFilter({...props.currentFilter, source_names})
+              }
+            />
+          ): (
+            <Input
+              placeholder="regular experssion"
```
Comment (r0rshark) [inline sapp/ui/frontend/src/Filter.js:238]:
```suggestion
              placeholder="regular expression"
```

Comment (r0rshark) [reply sapp/ui/frontend/src/Filter.js:238]: Can you fix also the other instances?

**Review by r0rshark — CHANGES_REQUESTED** [review]
Besides the typo it looks good to me

File: sapp/ui/frontend/src/Filter.js (hunk 1 above)
Comment (abishekvashok) [reply sapp/ui/frontend/src/Filter.js:238]: Sure

Comment (abishekvashok) [reply sapp/ui/frontend/src/Filter.js]: @r0rshark done

**Comment by r0rshark** [comment]
@abishekvashok  can you check why black is complaining?

**Comment by abishekvashok** [comment]
@r0rshark can affirm black is not failing due to the changes - I haven't touched any py file in the diff. I can fix or supress it however :)

File: sapp/ui/frontend/src/Filter.js [hunk 2]
```diff
@@ -251,17 +290,50 @@ const SourceKinds = (
...
+                  title={(props.currentFilter.source_kinds || []).join('\n')}
+                  placement="bottom">
+                  <Text type="secondary" size="small">
+                    {(props.currentFilter.source_kinds || []).length}
+                  </Text>
+                </Tooltip>
+              }
+            />
+          )}
+        </Col>
```
Comment (dkgi) [inline sapp/ui/frontend/src/Filter.js:336]: Looks like we repeat this pattern a couple of times. Do you think this could be factored out into its own component?

**Review by dkgi — CHANGES_REQUESTED** [review]
This look good to me. Back for question.

File: sapp/ui/frontend/src/Filter.js (hunk 2 above)
Comment (abishekvashok) [reply sapp/ui/frontend/src/Filter.js:336]: I will try!

**Comment by abishekvashok** [comment]
@dkgi done :)

**Comment by r0rshark** [comment]
Looks good I wonder why the ESLint CI seems to have run on the previous version and not on the last one

**Comment by abishekvashok** [comment]
@r0rshark should be good now :)

File: sapp/ui/frontend/src/Filter.js [hunk 3]
```diff
@@ -187,12 +187,76 @@ const Paths = (
   );
 };
 
+const MatchesIsField = (
+  props: $ReadOnly<{
+    allOptions: mixed,
```
Comment (dkgi) [inline sapp/ui/frontend/src/Filter.js:192]: Should this be `$ReadOnlyList<string>`?

File: sapp/ui/frontend/src/Filter.js [hunk 4]
```diff
@@ -187,12 +187,76 @@ const Paths = (
...
+  );
+};
+
 const SourceNames = (
   props: $ReadOnly<{
     currentFilter: FilterDescription,
     setCurrentFilter: FilterDescription => void,
   }>,
 ): React$Node => {
+
```
Comment (dkgi) [inline sapp/ui/frontend/src/Filter.js:259]: Is the change here coming from the auto-formatter?

**Review by dkgi — CHANGES_REQUESTED** [review]
Looks good to me. Minor nit.

File: sapp/ui/frontend/src/Filter.js (hunk 4 above)
Comment (abishekvashok) [reply sapp/ui/frontend/src/Filter.js:259]: Yes, kinda added a wierd vim plugin. Will remove it.

File: sapp/ui/frontend/src/Filter.js (hunk 3 above)
Comment (abishekvashok) [reply sapp/ui/frontend/src/Filter.js:192]: That's a much better type description!

**Comment by gbleaney** [comment]
Does client side regex matching really make sense here? Shouldn't this be something we push down to the DB query? I don't want us to have to return the entire DB to the client just to find that one issues with a unique sink

**Comment by abishekvashok** [comment]
> Does client side regex matching really make sense here? Shouldn't this be something we push down to the DB query? I don't want us to have to return the entire DB to the client just to find that one issues with a unique sink

We are getting a list of sources relating to a run already to show options, just like every other option to filtering. We are matching with that set just like we do for callables!

**Comment by abishekvashok** [comment]
@dkgi the requested changes have been made. Thank you for your time :)

**Comment by r0rshark** [comment]
> ex matching really make sense here? Shouldn't this be something we push down to the DB query? I don't want us to have to return the entire DB to the client just to find that one issues with a unique sink

It's slightly different than client side filtering. The way it works is:
0. We fetch the possible "sink kind" values e.g. "ReturnedToUser", "XSS"
1. Clientside  we do regex matching between the input field and the possible "sink kinds" e.g. with Input value "Retur" -> will get a list containing  "ReturnedToUser"
2. We send a graphQL query is sink_kinds: ["ReturnedToUser"]

I agree this is still not optimal since we need to return to the client the list of all the possible "Sink Kinds"/Names but it's better than returning the entire db :)
This approach was initially implemented by @dkgi who may have more context 
Screenshot 2021-08-09 at 13 57 41

**Comment by r0rshark** [comment]
> ex matching really make sense here? Shouldn't this be something we push down to the DB query? I don't want us to have to return the entire DB to the client just to find that one issues with a unique sink

It's slightly different than client side filtering. The way it works is:
0. We fetch the possible "sink kind" values e.g. "ReturnedToUser", "XSS"
1. Clientside  we do regex matching between the input field and the possible "sink kinds" e.g. with Input value "Retur" -> will get a list containing  "ReturnedToUser"
2. We send a graphQL query is sink_kinds: ["ReturnedToUser"]

I agree this is still not optimal since we need to return to the client the list of all the possible "Sink Kinds"/Names but it's better than returning the entire db :)
This approach was initially implemented by @dkgi who may have more context 
Screenshot 2021-08-09 at 13 57 41

**Comment by 0xedward** [comment]
> 0. We fetch the possible "sink kind" values e.g. "ReturnedToUser", "XSS"
>1. Clientside we do regex matching between the input field and the possible "sink kinds" e.g. with Input value "Retur" -> will get a list containing "ReturnedToUser"
>2. We send a graphQL query is sink_kinds: ["ReturnedToUser"]

> Does client side regex matching really make sense here? Shouldn't this be something we push down to the DB query? I don't want us to have to return the entire DB to the client just to find that one issues with a unique sink

It seems possible to also do this by:
1. passing the regex from the UI to the flask server with a graphql query
2. extending SQLAlchemy `String` class (e.g. https://gist.github.com/Xion/204ddbd020f1a4275a53) 
3. adding something like `where_regex_matches` in [`Instance`](https://github.com/facebook/sapp/blob/4adba9589f3ae28dd265c5597fbf8b99d39e75a4/sapp/ui/issues.py#L223) to build the DB query
4. run the DB query built in step 3 and return the list of all sources/sinks that match the regex

**Comment by dkgi** [comment]
I think we originally did this on the client side because it was cheap enough (even sending all possible function names to the client wasn't a big deal). It also allows us to provide live feedback to the user what their current regex is matching which is nice. Downside here is that it's hard to share the logic, and could potentially suck if we're sending back a large set to the query...

Don't have strong opinions myself.

**Comment by abishekvashok** [comment]
I can change this if we want to do it the other way, but as @dkgi said, we already have a list of source names (and every other field with us on the frontend to show as suggestions). We pass the needed ones to the backend in the form of a `filterDescription` to specify a filter. A regex matching and filtering by regex hence shouldn't be a big deal. The passed filters would look like the same when passed without the regex. But, I can change this if we would like to make a switch as @0xedward mentions :)

**Comment by gbleaney** [comment]
Ahh, I didn't realize we were doing *all* regex filtering client side. I get that it's not a perf problem (right now), but it still does present other problems. My biggest concern is: doesn't this make CLI filtering incompatible with UI filtering? IE. I can create a filter in the UI and get it perfect using RegEx components, but be unable to run it from the command line. That's a problem, because we assume that when people are using Pysa in CI, they're going to want to run the command line filtering rather than start up a whole webserver and headless browser in their CI.

Assuming that that concern is correct, @r0rshark and @abishekvashok, I'm pretty strongly in favour of switching to server side evaluation of the regex. I've got no problem with *also* offering live matching on the client side (assuming you can ensure you've got compatible regex engines running in both places) to make the UX better, but the canonical filtering should be done in the backend so CLI and UI filters are compatible.

**Comment by abishekvashok** [comment]
@gbleaney nice thought!
Currently, backend didn't have regex support - so without the sapp frontend regex isn't available via cli.
In the long term, I guess the changes pointed makes more sense, so I moved the regex matching to backend (also migrated existing filter parameters that had regex matching [callables] and their tests).

**Review by r0rshark — CHANGES_REQUESTED** [review]
Thanks for working on this! From my understanding with these changes we would be able to support creating a filter with regex filtering in the UI and using it in the commandline to filter issues. Can you validate this?
I just have a concern in the way we store the regex filtering information in the filter because having the type of match as the first element in the list look a bit hackish

File: sapp/ui/filter_predicates.py [hunk 5]
```diff
@@ -103,40 +114,36 @@ def __init__(
     ) -> None:
         self._parameter_list = parameter_list
         self._parameter_type = parameter_type
+        self._use_regex = (
+            True
+            if len(parameter_list) == 2 and "__regex_match__" in parameter_list
```
[the remark this answers is not in the document]
Comment (abishekvashok) [reply sapp/ui/filter_predicates.py:119]: Yeah, I added `"__regex_match__"` as the first element as I was trying a way to keep the existing filter layout! (Backward compatibility). I can change it to the mentioned layout, but I don't know whether existing filters will work :/

[the remark this answers is not in the document]
Comment (gbleaney) [reply sapp/ui/filter_predicates.py:119]: A good middle ground here is that you can check the type that you received. If you received a list, treat it as an `equals` operation and emit a warning saying that the user should upgrade their filter to the new format. If you receive a dict, parse as normal.

[the remark this answers is not in the document]
Comment (r0rshark) [reply sapp/ui/filter_predicates.py:119]: I agree with Graham, let's try this approach :)

[the remark this answers is not in the document]
Comment (abishekvashok) [reply sapp/ui/filter_predicates.py]: Done :)

**Comment by abishekvashok** [comment]
Fixing tests now :)

**Comment by abishekvashok** [comment]
@r0rshark @gbleaney thanks.

> From my understanding with these changes we would be able to support creating a filter with regex filtering in the UI and using it in the commandline to filter issues. Can you validate this?

As requested here's a testing of the backend.

https://user-images.githubusercontent.com/8947010/129366782-166ed985-2daf-4479-b27b-f92d4f2aeb46.mov

Also, I found out that we would need to export a filter created in the backend as a json file, and then use the json file to query. Potentially, I would love to make a feature request to be able to filter with just filter name on the backend (^^)

**Comment by 0xedward** [comment]
> Also, I found out that we would need to export a filter created in the backend as a json file, and then use the json file to query. Potentially, I would love to make a feature request to be able to filter with just filter name on the backend (^^)

@abishekvashok The current `filter issues` command was built for the upcoming SARIF support and was intended to be used in CI. 

I think that would be a good feature to add to our IPython interface ([`interactive.py`](https://github.com/facebook/sapp/blob/main/sapp/ui/interactive.py)) :)

File: sapp/ui/frontend/src/Filter.js [hunk 6]
```diff
@@ -187,6 +187,86 @@ const Paths = (
...
+        {selectedOperation === 'matches' ? (
+          <Input
+            placeholder="regular expression"
+            style={{width: '100%'}}
+            value={inputValue}
+            onChange={event => {
+              const value = event.target.value;
+              setInputValue(value);
+              const option_values = props.allOptions.filter(option =>
+                  option.match(value),
```
Comment (gbleaney) [inline sapp/ui/frontend/src/Filter.js:238]: Do Python and JavaScript use the same Regex engines? If not, trying to run the same regex on both might end up with confusing subtleties where frontend results don't match backend results

File: sapp/ui/issues.py [hunk 7]
```diff
@@ -462,26 +476,100 @@ def where_filter(self, filter_instance: Filter) -> "Instance":
             .where_status_is_any_of(filter_instance.statuses)
         )
 
+        if filter_instance.callables is not None and len(filter_instance.callables) > 0:
```
Comment (gbleaney) [inline sapp/ui/issues.py:479]: Both `None` and empty collections are falsey, so this is more concise
```suggestion
        if filter_instance.callables:
```

File: sapp/ui/issues.py [hunk 8]
```diff
@@ -462,26 +476,100 @@ def where_filter(self, filter_instance: Filter) -> "Instance":
...
+            if type(filter_instance.callables) is not list:
```
Comment (gbleaney) [inline sapp/ui/issues.py:480]: I'm confused my this. Above, `callables` is typed as a `List[Dict[...`, and before it was a `List[str]`. Shouldn't this branch never be hit? Based on the code below, it looks like you expect it to be a `Dict`. If that's the case, I think you need to update the typing on `Filter`

File: sapp/ui/issues.py [hunk 9]
```diff
@@ -462,26 +476,100 @@ def where_filter(self, filter_instance: Filter) -> "Instance":
...
+                if filter_instance.callables.get("operation") == "matches":
+                    builder = builder.where_callables_matches(
+                        filter_instance.callables.get("value", [""])[0]
+                    )
+                elif len(filter_instance.callables.get("value", [])) > 0:
+                    builder = builder.where_callables_is_any_of(
+                        filter_instance.callables.get("value", ["%"])
```
Comment (gbleaney) [inline sapp/ui/issues.py:487]: We have a default value of `["%"]` here, but nowhere else. I get that you're propagating the default value of `%` that you deleted from the `Query` object below, but I think that default was needed due to the unconditional `where_callables_is_any_of` that was being done above. Now that you've removed that, we'll only run the `where_callables_is_any_of` check if we actually have something to filter on, so I don't think the default is needed any more.

File: sapp/ui/tests/issues_test.py [hunk 10]
```diff
@@ -150,8 +150,10 @@ def testWhereCallables(self) -> None:
             builder = Instance(session, latest_run_id)
             issue_ids = {
                 int(issue.issue_instance_id)
-                for issue in builder.where_callables_is_any_of(["%function3"])
-                .where_callables_is_any_of(["%sub%"])
+                for issue in builder.where_callables_is_any_of(
+                    ["__regex_match__", ".*function3"]
+                )
+                .where_callables_is_any_of(["__regex_match__", ".*sub.*"])
```
Comment (gbleaney) [inline sapp/ui/tests/issues_test.py:156]: Out of date test?

File: sapp/ui/tests/issues_test.py [hunk 11]
```diff
@@ -150,8 +150,10 @@ def testWhereCallables(self) -> None:
             builder = Instance(session, latest_run_id)
             issue_ids = {
                 int(issue.issue_instance_id)
-                for issue in builder.where_callables_is_any_of(["%function3"])
```
Comment (gbleaney) [inline sapp/ui/tests/issues_test.py:153]: Could you please add tests for the source and sink filtering you added in this PR?

**Review by gbleaney — CHANGES_REQUESTED** [review]
Thanks for the big refactor! Looks pretty good overall, just some small concerns to address

File: sapp/ui/issues.py [hunk 12]
```diff
@@ -462,26 +476,100 @@ def where_filter(self, filter_instance: Filter) -> "Instance":
             .where_status_is_any_of(filter_instance.statuses)
         )
 
+        if filter_instance.callables is not None and len(filter_instance.callables) > 0:
+            if type(filter_instance.callables) is not list:
+                if filter_instance.callables.get("operation") == "matches":
+                    builder = builder.where_callables_matches(
+                        filter_instance.callables.get("value", [""])[0]
+                    )
+                elif len(filter_instance.callables.get("value", [])) > 0:
```
[the remark this answers is not in the document]
Comment (abishekvashok) [reply sapp/ui/issues.py:485]: I guess that makes the else hit on filters with empty callables such as:
```json
{
operation: "is",
value: ""
}
```
and give out no issues! Is that intended?

File: sapp/ui/issues.py (hunk 9 above)
Comment (abishekvashok) [reply sapp/ui/issues.py:487]: You are right! I was being too defensive (code) there!

File: sapp/ui/filter_predicates.py [hunk 13]
```diff
@@ -97,46 +108,42 @@ def apply(self, issues: List[IssueQueryResult]) -> List[IssueQueryResult]:
...
+        attribute = issue._asdict()[self._parameter_name]
+        if type(attribute) is str:
+            return {attribute}
+        return set(attribute)
+
+    def apply(self, issues: List[IssueQueryResult]) -> List[IssueQueryResult]:
+        return [
+            issue
+            for issue in issues
+            if len(list(filter(self._regex.match, self.attribute_set(issue)))) > 0
```
[the remark this answers is not in the document]
Comment (abishekvashok) [reply sapp/ui/filter_predicates.py]: Really insightful! Thanks, it makes me happy as well :)

File: sapp/ui/frontend/src/Filter.js (hunk 6 above)
Comment (abishekvashok) [reply sapp/ui/frontend/src/Filter.js:237]: I never thought we would have large differences, but it seems the case! So I updated the backend to use a library called js_regex!

File: sapp/ui/issues.py (hunk 8 above)
Comment (abishekvashok) [reply sapp/ui/issues.py:480]: It was actually to be `Dict`. Updated the typing on Filter :)

File: sapp/ui/tests/issues_test.py (hunk 10 above)
Comment (abishekvashok) [reply sapp/ui/tests/issues_test.py]: Yeah :cheeky: missed that! Sorry.

File: sapp/ui/tests/issues_test.py (hunk 11 above)
Comment (abishekvashok) [reply sapp/ui/tests/issues_test.py]: We lacked tests for sources and sinks, added them :)

**Comment by abishekvashok** [comment]
@gbleaney thank you for the in-depth review. As someone who enjoys tiny optimisations, loved the suggestions and enjoyed learning them :)

Fixed everything as suggested except one - left a comment just to make sure :)
Again, thanks for going in-depth 😁

File: requirements.txt [hunk 14]
```diff
@@ -6,6 +6,7 @@ flask-cors
 graphene==2.1.3
 graphene_sqlalchemy
 ipython
+js-regex
```
Comment (0xedward) [inline requirements.txt:9]:
```suggestion

```

File: .pyre_configuration [hunk 15]
```diff
@@ -28,6 +28,9 @@
     {
       "site-package": "IPython"
     },
+    {
+      "site-package": "js_regex"
+    },
```
Comment (0xedward) [inline .pyre_configuration:33]:
```suggestion
```

**Review by 0xedward — CHANGES_REQUESTED** [review]
LGTM :) I had a suggestion to remove the `js-regex` package, since it seems like it isn't being used.

**Comment by abishekvashok** [comment]
@0xedward sorry I don't how I missed adding a file, but js-regex is used, it's me messing up - sorry :/

File: sapp/ui/filter_predicates.py [hunk 16]
```diff
@@ -5,8 +5,19 @@
 
 from __future__ import annotations
 
+import js_regex
```
Comment (arthaud) [inline sapp/ui/filter_predicates.py:8]: nitpick: since this is not in the standard library, I would put it in the next import block (next to sqlalchemy stuff)

File: sapp/ui/filter_predicates.py [hunk 17]
```diff
@@ -97,46 +108,42 @@ def apply(self, issues: List[IssueQueryResult]) -> List[IssueQueryResult]:
...
 
 
+class Matches(IssuePredicate):
+    def __init__(self, regex: str, parameter_name: str) -> None:
+        self._regex: Pattern[str] = js_regex.compile(regex)
+        self._parameter_name = parameter_name
+
+    def attribute_set(self, issue: IssueQueryResult) -> Set[str]:
+        attribute = issue._asdict()[self._parameter_name]
+        if type(attribute) is str:
```
Comment (arthaud) [inline sapp/ui/filter_predicates.py:118]: Here and other places: I would prefer `isinstance(attribute, str)` over `type(attribute) is str`.

File: sapp/ui/issues.py (hunk 8 above)
Comment (arthaud) [reply sapp/ui/issues.py:506]: ditto: would use `isinstance`

File: sapp/filter.py [hunk 18]
```diff
@@ -17,7 +17,9 @@
 from .models import Base
 
 if TYPE_CHECKING:
-    from .ui.schema import FeatureCondition
+    from .ui.schema import FeatureCondition, MatchesIsField
+
+MatchesIsAlias = Union[Dict[str, Union[str, List[str]]], List[str]]
```
Comment (arthaud) [inline sapp/filter.py:22]: I think we could use a `TypedDict` here, to better describe what that `Dict[str, Union[str, List[str]]]` looks like.

File: sapp/ui/issues.py [hunk 19]
```diff
@@ -488,26 +502,95 @@ def where_filter(self, filter_instance: Filter) -> "Instance":
...
-        ):
-            builder = builder.where_sink_name_is_any_of(filter_instance.sink_names)
-        if (
-            filter_instance.sink_kinds is not None
-            and len(filter_instance.sink_kinds) > 0
-        ):
-            builder = builder.where_sink_kind_is_any_of(filter_instance.sink_kinds)
+        if filter_instance.callables:
+            if type(filter_instance.callables) is not list:
+                # pyre-fixme[16]: `List` has no attribute 'get'
```
Comment (arthaud) [inline sapp/ui/issues.py:507]: Why is that necessary? pyre should be able to infer that we have a `Dict` here.

File: sapp/ui/issues.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (arthaud) [reply sapp/ui/issues.py:512]: I would have a check `elif filter_instance.callables.get("operation") == "is":` and a `else` that throws an error.

File: sapp/ui/issues.py [hunk 20]
```diff
@@ -488,26 +502,95 @@ def where_filter(self, filter_instance: Filter) -> "Instance":
...
+                    )
+                elif len(filter_instance.callables.get("value", [])) > 0:
+                    builder = builder.where_callables_is_any_of(
+                        filter_instance.callables.get("value")
+                    )
+            else:
+                # Backward compatibility
+                # pyre-fixme[6]: Expected `List[str]` for 1st positional only
+                # parameter but got
+                # `Union[Dict[str, Union[List[str], str]], List[str]]`
```
Comment (arthaud) [inline sapp/ui/issues.py:520]: ditto, surprised pyre cannot infer that we have a list here. Maybe because you use `type(..) is list` instead of `isinstance()`?

**Review by arthaud — CHANGES_REQUESTED** [review]

File: sapp/ui/filter_predicates.py (hunk 16 above)
Comment (abishekvashok) [reply sapp/ui/filter_predicates.py]: Yep moving it :)

File: sapp/ui/filter_predicates.py (hunk 17 above)
Comment (abishekvashok) [reply sapp/ui/filter_predicates.py]: Sure

File: sapp/ui/issues.py (hunk 20 above)
Comment (abishekvashok) [reply sapp/ui/issues.py:525]: Migrated to `isinstance()`. Still pyre fails!

File: sapp/ui/issues.py (hunk 19 above)
Comment (abishekvashok) [reply sapp/ui/issues.py:507]: No it  isn't.

File: sapp/filter.py (hunk 18 above)
Comment (abishekvashok) [reply sapp/filter.py]: Done

**Comment by abishekvashok** [comment]
@arthaud thanks for your time reviewing :) I tried all your suggestions and pushed everything except for removing the pyre-check fixme comments. Pyre for some reason can't infer the type. Maybe I am missing something, but not sure!

File: sapp/ui/filter_predicates.py [hunk 21]
```diff
@@ -6,8 +6,19 @@
...
+    List,
+    Optional,
+    Pattern,
+    Sequence,
+    Set,
+    TypeVar,
+    Union,
+)
+
+import js_regex
```
Comment (r0rshark) [inline sapp/ui/filter_predicates.py:21]: @abishekvashok Why do we need to import js_regex can't we use the regular python regexes? js_regex https://pypi.org/project/js-regex/ looks like it simply support the javascript regex syntax which is not really something we care about. Moreover this adds a bit of maintenance burden since we would need to support this dependency  internally. Can we use normal python regex instead :)

**Review by r0rshark — CHANGES_REQUESTED** [review]

File: sapp/ui/filter_predicates.py (hunk 21 above)
Comment (abishekvashok) [reply sapp/ui/filter_predicates.py:21]: @r0rshark yes we can. But there are differences between the regex engines used by javascript and python. I switched to js-regex following this comment, https://github.com/facebook/sapp/pull/38#discussion_r690458934

But yeah, I will make the change and remove `js-regex` if we don't care much about the numbers in the frontend :)

**Comment by abishekvashok** [comment]
@r0rshark done :)

**Comment by abishekvashok** [comment]
@r0rshark done :)
